compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/Canonicalization.cpp

```diff
@@ -1048,6 +1048,46 @@ struct ZeroExtentTensorCanon final : RewritePattern {
   }
 };
 
+struct ReorderElementwiseAndShapeOp final
+    : OpTraitRewritePattern<OpTrait::Elementwise> {
+  using OpTraitRewritePattern::OpTraitRewritePattern;
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (op->getOperands().size() != 1) {
+      return rewriter.notifyMatchFailure(op, "expected to be unary.");
+    }
+
+    auto definingOp = op->getOperand(0).getDefiningOp();
+    if (!definingOp) {
+      return rewriter.notifyMatchFailure(
+          op, "expected to have an op before elementise op.");
+    }
+
+    if (!isa<mlir::stablehlo::ReshapeOp>(definingOp) &&
+        !isa<mlir::stablehlo::TransposeOp>(definingOp) &&
+        !isa<mlir::stablehlo::BroadcastOp>(definingOp)) {
+      return rewriter.notifyMatchFailure(
+          op, "defining operation of unexpected type.");
+    }
+
+    Value input = definingOp->getOperand(0);
+    Value result = op->getResult(0);
+    if (getElementTypeOrSelf(input) != getElementTypeOrSelf(result)) {
```

Comment:
You should be able to generate a new intermediate type by `result.getType().clone(getElementTypeOrSelf(input.getType()))`

Reply:
Done (but the other way around -- this version is already the intermediate type :)